### PR TITLE
Getting accounts for private urls

### PR DESCRIPTION
**Pull request by dhoehna**
## Summary of the pull request
Fixes a few issues.
1. First and foremost, prompts the user to log in if they provided a url to a private repo.
2. Auto-advances the account page to the repo list after the user logs into their account.
3. Moved some logic from AddRepoView to AddRepoViewModel and added a test.

## Detailed description of the pull request / Additional comments
One fundamental with url based cloning is both a private repo, and a repo that does not exist, will return a 404.  Meaning DevHome does not have a way to recognize when a url needs a user to log in.  This means DevHome will ask the user to log in for both a private repo and one that does not exist.  

## Validation steps performed
Ran the login scenario for the account page and URl page for private repos.

## New ui bug introduced by this PR
Because the url login logic is in the primary button handler, the UI won't handle the close button.  To remedy this, I added a close button to the grid that display the login UI when logging in on the URL screen.

If you notice, the "X" button does not go away when the user clicks the login button.  This is because DevHome has no way to listen to this button click.  DevHome does its best and waits for login confirmation from the developer id provider.

A few ways to fix this.
1. Move the login URL logic to the view model then, if the user logged in add the repo to EverythingToClone.  Somehow call .Hide().  Then, in the RepoConfigDialog check if the number of repos to clone increased.  If so, a repo was added.
2. Move the login URL logic to the view model.  After a successful user login somehow tell the ContentDialog to execute the primary button click to add the repo.

## PR checklist
- [ ] Closes #xxx
- [ ] Tests added/passed
- [ ] Documentation updated

Logging in via URL!
LoggingInViaUrl

Logging in via Account!
LoggingInViaAccount

### CODE REVIEW

File: extensionsdk/Microsoft.Windows.DevHome.SDK/Microsoft.Windows.DevHome.SDK.idl [hunk 1]
```diff
@@ -226,7 +226,7 @@ namespace Microsoft.Windows.DevHome.SDK
 
         ProviderOperationResult LogoutDeveloperId(IDeveloperId developerId);
 
-        event Windows.Foundation.TypedEventHandler<IDeveloperIdProvider, IDeveloperId> Changed;
+        event Windows.Foundation.TypedEventHandler<IDeveloperIdProvider> Changed;
```
Comment (vineeththomasalex) [inline extensionsdk/Microsoft.Windows.DevHome.SDK/Microsoft.Windows.DevHome.SDK.idl:229]: This was likely an unintentional change?

File: tools/SetupFlow/DevHome.SetupFlow/Views/AddRepoDialog.xaml.cs [hunk 2]
```diff
@@ -96,6 +99,36 @@ public async Task GetExtensionsAsync()
         await Task.Run(() => AddRepoViewModel.GetExtensions());
     }
 
+    /// <summary>
+    /// Sets the event handler on all providers to listen when the user has logged in.
+    /// </summary>
+    public void SetDeveloperChangedEvents()
```
Comment (vineeththomasalex) [inline tools/SetupFlow/DevHome.SetupFlow/Views/AddRepoDialog.xaml.cs:105]: Nit: SetDeveloper**Id**ChangedEvents

File: tools/SetupFlow/DevHome.SetupFlow/Views/AddRepoDialog.xaml.cs [hunk 3]
```diff
@@ -96,6 +99,36 @@ public async Task GetExtensionsAsync()
...
+    {
+        AddRepoViewModel.SetChangedEvents(ChangedEventHandler);
+    }
+
+    /// <summary>
+    /// Changes the flag that indicates if the user is logging in to false to indicate the user is done logging in.
+    /// </summary>
+    /// <param name="sender">The object that raised this event, should only be IDeveloperId</param>
+    /// <param name="developerId">The developer the log in action is applied to.</param>
+    public void ChangedEventHandler(object sender, IDeveloperId developerId)
```
Comment (vineeththomasalex) [inline tools/SetupFlow/DevHome.SetupFlow/Views/AddRepoDialog.xaml.cs:115]: Nit: **DeveloperId**ChangedEventHandler

File: tools/SetupFlow/DevHome.SetupFlow/Models/RepositoryProvider.cs [hunk 4]
```diff
@@ -108,50 +127,23 @@ public IRepository GetRepositoryFromUri(Uri uri, IDeveloperId developerId = null
     /// Checks with the provider if it understands and can clone a repo via Uri.
     /// </summary>
     /// <param name="uri">The uri to the repository</param>
-    /// <returns>A tuple that containes if the provider can parse the uri and the account it can parse with.</returns>
-    /// <remarks>If the provider can't parse the Uri, this will try a second time with any logged in accounts.  If the repo is
-    /// public, the developerid can be null.</remarks>
-    public (bool, IDeveloperId, IRepositoryProvider) IsUriSupported(Uri uri)
+    /// <returns>True if this provider supports the url.  False otherwise.</returns>
+    public bool IsUriSupported(Uri uri)
```
Comment (vineeththomasalex) [inline tools/SetupFlow/DevHome.SetupFlow/Models/RepositoryProvider.cs:131]: Nit: IsUriSupportedByAnyDeveloperId()
Or maybe clarifying with a comment is enough

File: extensionsdk/Microsoft.Windows.DevHome.SDK/Microsoft.Windows.DevHome.SDK.idl (hunk 1 above)
Comment (dhoehna) [reply extensionsdk/Microsoft.Windows.DevHome.SDK/Microsoft.Windows.DevHome.SDK.idl]: Yep.  Changed back.

File: tools/SetupFlow/DevHome.SetupFlow/Views/AddRepoDialog.xaml.cs (hunk 2 above)
Comment (dhoehna) [reply tools/SetupFlow/DevHome.SetupFlow/Views/AddRepoDialog.xaml.cs:106]: Changed.

File: tools/SetupFlow/DevHome.SetupFlow/Views/AddRepoDialog.xaml.cs (hunk 3 above)
Comment (dhoehna) [reply tools/SetupFlow/DevHome.SetupFlow/Views/AddRepoDialog.xaml.cs:116]: Changed.

File: tools/SetupFlow/DevHome.SetupFlow/Models/RepositoryProvider.cs (hunk 4 above)
Comment (dhoehna) [reply tools/SetupFlow/DevHome.SetupFlow/Models/RepositoryProvider.cs:131]: Added comments explaining the method.

File: tools/SetupFlow/DevHome.SetupFlow/ViewModels/AddRepoViewModel.cs [hunk 5]
```diff
@@ -581,6 +630,120 @@ public void AddRepositoryViaUri(string url, string cloneLocation)
...
+                    cloningInformation.CloningLocation = new DirectoryInfo(cloneLocation);
+                    cloningInformation.OwningAccount = loggedInAccount;
+
+                    return cloningInformation;
+                }
+            }
+
+            // In the case that no logged in accounts can access it, return null
+            // until DevHome can handle multiple accounts.
+            // Should have a better error string.
```
Comment (krschau) [inline tools/SetupFlow/DevHome.SetupFlow/ViewModels/AddRepoViewModel.cs:691]: Nit: use word "todo" so it's findable as a todo.

File: tools/SetupFlow/DevHome.SetupFlow/Models/RepoViewListItem.cs [hunk 6]
```diff
@@ -1,6 +1,7 @@
 ﻿// Copyright (c) Microsoft Corporation and Contributors
 // Licensed under the MIT license.
 
+using System.Collections.ObjectModel;
```
[the remark this answers is not in the document]
Comment (dhoehna) [reply tools/SetupFlow/DevHome.SetupFlow/Models/RepoViewListItem.cs:4]: Most likely an artifact of early POC testing.  Removed.

File: tools/Dashboard/DevHome.Dashboard/Views/WidgetViewModel.cs
[the remark this answers is not in the document]
Comment (dhoehna) [reply tools/Dashboard/DevHome.Dashboard/Views/WidgetViewModel.cs:1]: I have no idea how that happened.  Changed back.

File: tools/SetupFlow/DevHome.SetupFlow/Services/StringResourceKey.cs [hunk 7]
```diff
@@ -1,6 +1,8 @@
 ﻿// Copyright (c) Microsoft Corporation and Contributors
 // Licensed under the MIT license.
 
+using Windows.Media.Streaming.Adaptive;
+
```
[the remark this answers is not in the document]
Comment (dhoehna) [reply tools/SetupFlow/DevHome.SetupFlow/Services/StringResourceKey.cs:5]: Odd.  Removed.

File: tools/SetupFlow/DevHome.SetupFlow/Strings/en-us/Resources.resw [hunk 8]
```diff
@@ -1235,4 +1235,12 @@
     <value>Starting to install package {0}</value>
     <comment>{Locked="{0}"} String letting users know that DevHome has started to install a package. {0} is the package id</comment>
   </data>
+  <data name="UrlCancelButtonText" xml:space="preserve">
+    <value>Press the X button to cancel.</value>
```
[the remark this answers is not in the document]
Comment (dhoehna) [reply tools/SetupFlow/DevHome.SetupFlow/Strings/en-us/Resources.resw:1239]: I don't understand.  A "real" string?  This isn't specific to github.

File: tools/SetupFlow/DevHome.SetupFlow/ViewModels/AddRepoViewModel.cs [hunk 9]
```diff
@@ -500,73 +561,61 @@ public void AddRepositoryViaUri(string url, string cloneLocation)
...
-                var uriBuilder = new UriBuilder(parsedUri.OriginalString);
+                var uriBuilder = new UriBuilder(uri.OriginalString);
                 uriBuilder.Port = -1;
-                parsedUri = uriBuilder.Uri;
+                uri = uriBuilder.Uri;
             }
             catch (Exception e)
             {
-                Log.Logger?.ReportError(Log.Component.RepoConfig, $"Invalid URL {parsedUri.OriginalString}", e);
+                Log.Logger?.ReportError(Log.Component.RepoConfig, $"Invalid URL {uri.OriginalString}", e);
```
[the remark this answers is not in the document]
Comment (dhoehna) [reply tools/SetupFlow/DevHome.SetupFlow/ViewModels/AddRepoViewModel.cs:574]: `Uri.TryCreate` catches the instance the passed in string isn't a uri.  By the time this code hits, uri.OriginalString is not null or empty.

What problems are you anticipating?

File: tools/SetupFlow/DevHome.SetupFlow/ViewModels/AddRepoViewModel.cs (hunk 5 above)
Comment (dhoehna) [reply tools/SetupFlow/DevHome.SetupFlow/ViewModels/AddRepoViewModel.cs:691]: Added.

**Review by vineeththomasalex — APPROVED** [review]

File: tools/SetupFlow/DevHome.SetupFlow/Strings/en-us/Resources.resw (hunk 8 above)
[the remark this answers is not in the document]
Comment (krschau) [reply tools/SetupFlow/DevHome.SetupFlow/Strings/en-us/Resources.resw:1239]: Sorry, my comment was wildly unhelpful. I should have said "has the official copy writer person given us this string"? Something about calling it the "x" button seemed non-official but I have no idea.

File: tools/SetupFlow/DevHome.SetupFlow/ViewModels/AddRepoViewModel.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (krschau) [reply tools/SetupFlow/DevHome.SetupFlow/ViewModels/AddRepoViewModel.cs:574]: Just wondering.

**Review by krschau — APPROVED** [review]

**Pull request merged**